sources/api/shibaken/src/warmpool/autoscaling_warm_pool.rs

```diff
@@ -0,0 +1,60 @@
+/// This module contains utilities for fetching to IMDS to check if the
+/// instance on autoscaling warm pool is InService.
+use crate::warmpool::config::Config;
+use crate::warmpool::error::{self, Result};
+use argh::FromArgs;
+use imdsclient::ImdsClient;
+use log::info;
+use snafu::{OptionExt, ResultExt};
+use tokio::time::{sleep, Duration};
+
+// Path to config file containing host's autoscaling.should-wait setting
+const CONFIG_PATH: &str = "/etc/autoscaling-warm-pool-k8s.toml";
+const FETCH_LIFECYCLE_INTERVAL: u64 = 5;
+
+#[derive(FromArgs, Debug)]
+#[argh(subcommand, name = "is-autoscaling-warm-pool-healthy")]
+/// fetch to IMDS to check if the instance on autoscaling warm pool is InService
+pub(crate) struct IsAutoscalingWarmPoolHealthy {}
+
+impl IsAutoscalingWarmPoolHealthy {
+    pub(crate) async fn run(self) -> Result<()> {
+        wait_until_inservice().await?;
+        Ok(())
+    }
+}
+
+/// Uses an imdsclient function to fetch the lifecyclestate of the host
+async fn get_lifecycle_state(client: &mut ImdsClient) -> Result<String> {
+    client
+        .fetch_lifecycle_state()
+        .await
+        .context(error::ImdsRequestSnafu)?
+        .context(error::ImdsNoneSnafu {
+            what: "instance-id",
+        })
+}
+
+/// Continuously fetches lifecycle state of the host.
+/// Returns from function when host is "InService"
+async fn wait_until_inservice() -> Result<()> {
+    let mut client = ImdsClient::new();
+    let mut lifecycle_state = get_lifecycle_state(&mut client).await?;
+
+    info!("Lifecycle state is {}", lifecycle_state);
+
+    while lifecycle_state.ne("InService") {
+        lifecycle_state = get_lifecycle_state(&mut client).await?;
+        info!("Lifecycle state is {} ... waiting.", lifecycle_state);
+        sleep(Duration::from_secs(FETCH_LIFECYCLE_INTERVAL)).await;
+    }
```

Comment:
This imposes an overhead of a five second wait even if the instance is already in service.